snuba_sdk/legacy.py

```diff
@@ -107,23 +184,17 @@ def parse_exp(value: Any) -> Any:
     elif not isinstance(value, list):
         return parse_scalar(value)
 
-    alias = value[2] if len(value) > 2 else None
-    if alias and alias.startswith("`") and alias.endswith("`"):
-        alias = alias[1:-1]
+    return parse_function(value)
 
-    if value[0].endswith("()") and not value[1]:
-        return Function(value[0].strip("()"), [], alias)
-    if not value[0].endswith(")") and not value[1]:
-        # ["count", None, "count"]
-        return Function(value[0], [], alias)
 
-    children = None
-    if isinstance(value[1], list):
-        children = list(map(parse_exp, value[1]))
-    elif value[1]:
-        children = [parse_exp(value[1])]
+def parse(value: Any) -> Any:
+    if not isinstance(value, list):
+        return parse_exp(value)
 
-    return Function(value[0], children, alias)
+    try:
+        return parse_exp(value)
+    except Exception:
+        return parse_legacy_function(value)
```

Comment:
Which exception do you expect here ?

Reply:
The specific error I was dealing with threw an IndexError, but I figured that really any error should be cause to try the legacy parser.

Reply:
Why don't we only run the legacy parser? 
If the old language behaved in that ambiguous way and this is meant to translate legacy queries into SnQL queries, I think we should just run the old parsing logic here. 
Relying on an exception to assume the issue is that we are using the wrong parser seems a very fragile and ad-hoc solution that would shadow real issues that manifest themselves in the same way by raising an Exception.
It also seems a smell that the parser relies on a function to raise an exception to do its normal behavior.

If the production environment depends on the non ambiguous syntax to translate legacy queries to SnQL query it seems to me we are doing something wrong.

Reply:
I agree this solution isn't ideal. I originally had the same idea you did, which was to solely use the logic from the legacy Snuba parser. However a large number of my tests started failing when I did that, with the resulting SnQL being inconsistent with the previous parsing. I am worried that if I change all this parsing now, I'm going to introduce bugs that weren't there before and will be harder to track since I removed the dry run SQL comparison.

Reply:
The more I think about this, the more I realize this: This change is purely to support a single Sentry test. It has not occurred in production yet. I'm going to try to fix the test to be unambiguous and see if this is actually something that could occur in production or is unique to the test.